plugins/index.ts

```diff
@@ -179,23 +179,44 @@ extendEnvironment((hre) => {
       hre.log(
         'Found existing proxy at:',
         proxyAddress,
-        ' attempting to upgrade instance',
+        'attempting to upgrade instance',
         contractName
       );
+      const existingImplementationAddress =
+        await hre.upgrades.erc1967.getImplementationAddress(proxyAddress!);
+      hre.log('Existing implementation at:', existingImplementationAddress);
       const fireblocksSigner = signer as FireblocksSigner;
       if (typeof fireblocksSigner.setNextTransactionMemo === 'function') {
-          fireblocksSigner.setNextTransactionMemo(`Upgrade contract instance for ${contractName}`);
+        fireblocksSigner.setNextTransactionMemo(
+          `Upgrade contract instance for ${contractName}`
+        );
+      }
+      const deployment = await hre.deployments.get(contractName);
+      const artifact = await hre.deployments.getArtifact(contractName);
+      if (deployment.bytecode !== artifact.bytecode) {
+        contract = await hre.upgrades.upgradeProxy<TContract>(
+          proxyAddress!,
+          contractFactory
+          // options
+        );
+        const newImplementationAddress =
+          await hre.upgrades.erc1967.getImplementationAddress(proxyAddress!);
+        if (existingImplementationAddress === newImplementationAddress) {
+          hre.log('Implementation unchanged');
+        } else {
+          hre.log('New implementation at:', newImplementationAddress);
+        }
+        hre.trace('...awaiting deployment transaction', contractName);
+        await contract.deployed();
+        hre.log('...successful deployment transaction', contractName);
+      } else {
+        hre.log('Implementation appears unchanged, skipped upgrade attempt.');
+        contract = (await getContract({
+          contractName,
+          hre,
+        })) as InstanceOfContract<TContract>;
```

Comment:
we should just make `getContract` take a generic argument

Reply:
Yeah, I gave that a meek attempt and moved on but I'd love your help to figure that out.